rfcs/0011-separate-flyway-from-pipeline.md

```diff
@@ -0,0 +1,382 @@
+building in Postgres by specifying the CONCURRENT keyword. Care must be given to the fact that the application will be
+online while the operation is running (so dropping an index that the application needs while building a new one
+asynchronously is not acceptable).
+
+Long-running synchronous migrations are the least desirable and require the most care. Examples include creating
+a new table as a copy of an existing large table, altering the datatype of a column in a large table, creating an
+index in a large table without specifying the CONCURRENT keyword. Due to the Flyway migration
+being performed during the startup of the BFD Pipeline application, a long-running synchronous migration that is
+deployed with a Jenkins deployment will block the Pipeline application from consuming data for the duration of the
+migration. The impact of this limitation will become more significant as the pipeline RDA job goes into production.
+
+This RFC proposal addresses the concern of long-running synchronous migrations blocking the BFD Pipeline as well as
+improvements to the Jenkins deployment pipeline to better support these long-running migrations when they are
+necessary.
+
+#### Other shortcomings that can be addressed by this proposal
+
+In addition to optimizing the development and deployment of database migrations, the proposed changes can also solve
+other issues related to database migrations:
+
+* Database object privileges and high level roles (not individual user roles) are not set consistently or in a way that 
+supports database credential rotations. Ideally these privileges and roles would be developed and then deployed across
+environments in a Flyway migration so that their creation is automated and embedded in the source code. This cannot
+currently be done because the Flyway invocation is part of the BFD Pipeline Application startup which must run as a
+specific role that is appropriate for the pipeline but does not have the privileges to create roles or set privileges on
+existing objects.
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+The proposed solution is to remove the invocations of Flyway and Hibernate validation from the current BFD applications
+and instead run Flyway and Hibernate validation as a step in the deployment that must complete successfully before
+continuing with the deployment of the BFD Pipeline application and BFD FHIR Server.
+
+By moving the Flyway migrations to a step before all other applications are deployed, it no longer is necessary to
+deploy a schema change separately from its accompanying application change or have forward-compatible changes (since the
+new application will only ever start up after the new schema has been deployed). Backward-compatibility remains a
+requirement because the old applications will still coexist with the new schema for a brief period of time.
+
```

Comment:
Thinking of the recent MbiCache PR sequence:

1. SQL only PR to create MbiCache table and add foreign key columns to claims tables
2. Combined PR with sql to populate new table from existing database records plus pipeline code update to start populating the new table and columns for newly arriving records
3. Code only PR that updates api to query using the new table and column instead of old columns and changes pipeline to stop updating the old columns.
4. SQL only PR to drop the old columns

I'm not sure how well this new process would work for this sort of migration.  The first sql only migration could be combined with the pipeline code changes so that's a win.  But when could the sql to populate the new table from existing records be executed?  It needs for the old pipeline to be stopped before it can reliably do that.  In this proposal the old pipeline will still be running while the migration is running.  This could even create conflicts between simultaneous updates to records by both the migration script and the pipeline app.

Also the cleanup migration would still have to take place in its own PR.  So we can't reliably run the sql from PR 2 and still need 3 and 4 to be separate PRs.

Would it be possible to:

1. Stop the old pipeline before kicking off the migrator app.
2. Add support for a POST deployment migration that could perform cleanup changes that only make sense once the new code is fulling running.

With these changes the old 4-PR sequence could be just 2 PRs (1+2 & 3+4 from the old sequence).

Reply:
Thanks for giving this a read.

We could fairly easily stop the pipeline application before running the migrator app but it means that the pipeline would not ingest data during the migration which could be 24 hours for some long running synchronous migrations that we envision happening. I was assuming that would be a non-starter for the RDA job.

There are ways to get around that though. We could make it an option of the deployment as to whether the pipeline application would be stopped or not during the deployment but that seems error prone to me. Another possibility would be to run that type of long running job in a post-deployment migration which you have also alluded to as another desirable feature. That feature is currently sitting in the Future Possibilities section and these are good use cases for it so I think we should keep it in mind. Luckily that would build on top of this proposal and not be in conflict with it.

I appreciate putting a real work change to the test but I think we should also consider whether this particular complex change is representative of typical migration changes that we expect to encounter most often. I believe the previous two that were done for PACA where we followed the BFD process to the letter (including the hibernate validation off/on steps for absolute safety even though we were not totally sure that they were needed) are real world examples of going from 4 PRs to 1.